README.md
# Rack::DiscordInteractions

Middleware to handle signature verification for Discord Interaction HTTP requests.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'rack-discord-interactions'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install rack-discord-interactions

## Usage

```ruby
require 'sinatra'
require 'rack/discord-interactions'
require 'json'

use Rack::DiscordInteractions::Verifier, ENV['INTERACTIONS_PUBLIC_KEY'], path: '/interactions'

post '/interactions' do
  body = JSON.parse(request.body.read)

  case body['type']
  when 2 # COMMAND
    { type: 5 }.to_json
  else
    halt 400, 'Invalid type'
  end
end
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/shardlab/rack-discord-interactions. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/shardlab/rack-discord-interactions/blob/master/CODE_OF_CONDUCT.md).


## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Rack::DiscordInteractions project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/shardlab/rack-discord-interactions/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2021-03-12
### Added
- Initial release

[Unreleased]: https://github.com/shardlab/rack-discord-interactions/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/shardlab/rack-discord-interactions/releases/tag/v1.0.0

spec/spec_helper.rb
# frozen_string_literal: true

require 'bundler/setup'
require 'rack/discord-interactions'

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = '.rspec_status'

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/rack/discord-interactions/verifier_spec.rb
# frozen_string_literal: true

require 'rack/discord-interactions'
require 'rack/mock'

RSpec.describe Rack::DiscordInteractions::Verifier do
  let(:app) { ->(env) { [200, env, []] } }
  let(:signing_key) { Ed25519::SigningKey.generate }
  let(:public_key) { signing_key.verify_key.to_bytes.unpack1('H*') }
  let(:verifier) do
    described_class.new(app, public_key, path: '/interactions')
  end

  describe '#initialize' do
    it 'converts the public key into a verifying key' do
      middleware_key = verifier.instance_variable_get(:@verify_key)

      expect(middleware_key.to_str).to eq signing_key.verify_key.to_str
    end
  end

  describe '#call' do
    context 'when the request matches the provided route' do
      it 'verifies the request' do
        allow(verifier).to receive(:verify)

        verifier.call(Rack::MockRequest.env_for('/interactions'))

        expect(verifier).to have_received(:verify).with an_instance_of(Rack::Request)
      end
    end

    context 'when the request does not match the provided route' do
      it 'does not verify the request' do
        allow(verifier).to receive(:verify)

        verifier.call(Rack::MockRequest.env_for('/not-interactions'))

        expect(verifier).not_to have_received(:verify)
      end

      it 'forwards the env to the app' do
        allow(app).to receive(:call)
        env = Rack::MockRequest.env_for('/not_interactions')

        verifier.call(env)

        expect(app).to have_received(:call).with env
      end
    end

    describe '#verify' do
      let(:request) { Rack::Request.new(Rack::MockRequest.env_for('/interactions')) }

      before do
        ts = Time.now.to_i.to_s
        data = '{"type": 2}'

        request.set_header('HTTP_X_SIGNATURE_TIMESTAMP', ts)
        request.set_header('HTTP_X_SIGNATURE_ED25519', signing_key.sign("#{ts}#{data}").unpack1('H*'))

        request.body.string = data
      end

      context 'when the timestamp header is missing' do
        before do
          request.delete_header('HTTP_X_SIGNATURE_TIMESTAMP')
        end

        it 'returns a bad request response' do
          response = Rack::MockResponse.new(*verifier.verify(request))

          expect(response.bad_request?).to be true
        end
      end

      context 'when the ed25519 header is missing' do
        it 'returns a bad request response' do
          request.delete_header('HTTP_X_SIGNATURE_ED25519')
          response = Rack::MockResponse.new(*verifier.verify(request))

          expect(response.bad_request?).to be true
        end
      end

      context 'when the signature is invalid' do
        it 'returns an unauthorized response' do
          request.set_header('HTTP_X_SIGNATURE_ED25519', 'invalid')
          response = Rack::MockResponse.new(*verifier.verify(request))

          expect(response.unauthorized?).to be true
        end
      end

      context 'when the signature does not match' do
        before do
          request.set_header('HTTP_X_SIGNATURE_ED25519', signing_key.sign('invalid').unpack1('H*'))
        end

        it 'returns an unauthorized response' do
          response = Rack::MockResponse.new(*verifier.verify(request))

          expect(response.unauthorized?).to be true
        end
      end

      context 'when the signature matches' do
        it 'passes the env to the app' do
          allow(app).to receive(:call)

          verifier.verify(request)

          expect(app).to have_received(:call).with(request.env)
        end
      end

      context 'when it is a ping' do
        before do
          allow(app).to receive(:call)
          ts = Time.now.to_i.to_s
          data = '{"type": 1}'

          request.set_header('HTTP_X_SIGNATURE_TIMESTAMP', ts)
          request.set_header('HTTP_X_SIGNATURE_ED25519', signing_key.sign("#{ts}#{data}").unpack1('H*'))

          request.body.string = data
        end

        it 'automatically responds to pings' do
          response = Rack::MockResponse.new(*verifier.verify(request))
          expect(JSON.parse(response.body)['type']).to eq 1
        end
      end

      context 'when handle_pings is false' do
        let(:verifier) do
          described_class.new(app, public_key, path: '/interactions', handle_pings: false)
        end

        before do
          allow(app).to receive(:call)
          ts = Time.now.to_i.to_s
          data = '{"type": 1}'

          request.set_header('HTTP_X_SIGNATURE_TIMESTAMP', ts)
          request.set_header('HTTP_X_SIGNATURE_ED25519', signing_key.sign("#{ts}#{data}").unpack1('H*'))

          request.body.string = data
        end

        it 'passes the env to the app' do
          verifier.verify(request)

          expect(app).to have_received(:call).with(request.env)
        end
      end

      it 'verifies the signature' do
        response = Rack::MockResponse.new(*verifier.verify(request))
        expect(response.successful?).to be true
      end
    end
  end
end


lib/rack/discord-interactions/verifier.rb
# frozen_string_literal: true

require 'ed25519'
require 'json'

module Rack
  # Discord interaction middleware collection
  module DiscordInteractions
    # Middleware to handle verification of interaction signatures
    class Verifier
      # @!visibility private
      MISSING_HEADERS = [400, {}, ['Missing headers']].freeze

      # @!visibility private
      PONG = [200, { 'Content-Type' => 'application/json' }, ['{"type":1}']].freeze

      # @!visibility private
      INVALID_SIGNATURE = [401, {}, ['Invalid request signature']].freeze

      # @param app [#call] The rack app, or an object that responds to `#call`.
      # @param public_key [String] The public key of the application.
      # @param path [String] Path to verify, will verify all requests if `nil`.
      # @param handle_pings [true, false] Whether the middleware should automatically handle responding to pings.
      def initialize(app, public_key, path: nil, handle_pings: true)
        @app = app
        @verify_key = Ed25519::VerifyKey.new([public_key].pack('H*'))
        @path = path
        @handle_pings = handle_pings
      end

      # @!visibility private
      # @param env [Hash]
      def call(env)
        if env['PATH_INFO'] == @path || @path.nil?
          request = Rack::Request.new(env)
          verify(request)
        else
          @app.call(env)
        end
      end

      # @!visibility private
      # @param request [Rack::Request]
      def verify(request)
        timestamp = request.get_header('HTTP_X_SIGNATURE_TIMESTAMP')
        signature = request.get_header('HTTP_X_SIGNATURE_ED25519')
        body = request.body.read

        return MISSING_HEADERS unless timestamp && signature

        @verify_key.verify([signature].pack('H*'), timestamp + body)
        request.body.rewind

        return PONG if @handle_pings && JSON.parse(body)['type'] == 1

        @app.call(request.env)
      rescue ArgumentError, Ed25519::VerifyError
        INVALID_SIGNATURE
      end
    end
  end
end
